add asynchronous metrics for total primary key size in memory and allocated

bharatnc opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add asynchronous metrics for total primary key bytes and total allocated primary key bytes in memory.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

Add two new metrics `TotalPrimaryKeyBytesInMemory` and `TotalPrimaryKeyBytesInMemoryAllocated` to asynchronous metrics for the total size of primary key bytes and the total size of allocated primary key bytes in memory (only taking active parts into account). The logic for obtaining the size of primary key in memory for each part  is similar to how it's done for the `system.parts`  (`primary_key_bytes_in_memory` and `primary_key_bytes_in_memory_allocated`) and the values for all parts are summed to get a total value.
```
:) select metric, value from system.asynchronous_metrics WHERE metric LIKE '%TotalPrimaryKeyBytes%';

SELECT
    metric,
    value
FROM system.asynchronous_metrics
WHERE metric LIKE '%TotalPrimaryKeyBytes%'

Query id: 3f73fc81-8771-4a42-94a4-282ff6e7d247

┌─metric────────────────────────────────┬───value─┐
│ TotalPrimaryKeyBytesInMemory          │  187003 │
│ TotalPrimaryKeyBytesInMemoryAllocated │ 1887712 │
└───────────────────────────────────────┴─────────┘
```

Related to: #56779

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

bharatnc committed:
asynchronous metrics - total primary key bytes in memory

bharatnc committed:
add docs

bharatnc commented:
cc: @alexey-milovidov

robot-clickhouse commented:
<!-- automatic status comment for PR #57551 from bharatnc/ClickHouse:ncb/async-metrics-pri-key-size -->
*This is an automated comment for commit 41b2b63ceefa83fc2e7fb89fff6fa4641b2443a4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57551/41b2b63ceefa83fc2e7fb89fff6fa4641b2443a4/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7109653407">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57551/41b2b63ceefa83fc2e7fb89fff6fa4641b2443a4/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

bharatnc committed:
only account for active parts

bharatnc committed:
fix clang-tidy

bharatnc commented:
Probably need to add a test for these two new metrics, but not exactly sure what kind of test.

CheSema commented on src/Interpreters/ServerAsynchronousMetrics.cpp:
```diff
@@ -287,6 +290,17 @@ void ServerAsynchronousMetrics::updateImpl(AsynchronousMetricValues & new_values
                     }
+
+                    // fetch the list of all parts regardless of their state
+                    auto all_parts = table_merge_tree->getAllDataPartsVector();
```
As a person who looks at the metrics regularly, want to tell you thank you for this PR.

I see that 
`TotalPrimaryKeyBytesInMemory`, `TotalPrimaryKeyBytesInMemoryAllocated` - that naming goes along with totalBytes and other `Total` columns. It is OK.

In most cases when I need to know memory consumption by loaded indexes it includes `outdated` parts also.
I'm not sure about it, but how about add
`TotalConsumptionByPrimaryKeyBytesInMemory` -- sum(primary_key_bytes_in_memory_allocated) for all parts.

CheSema commented:
> Probably need to add a test for these two new metrics, but not exactly sure what kind of test.

For testing it you need an '.sh' tests in `ClickHouse/tests/queries/0_stateless/`.
Make some queries, write some data, delete some part (make sure that it stays as outdated for a while).
After that ask `asynchronous_metrics` as you did above. You may need to use `query_with_retry` du to asynchronous.

bharatnc commented:
> > Probably need to add a test for these two new metrics, but not exactly sure what kind of test.
> 
> For testing it you need an '.sh' tests in `ClickHouse/tests/queries/0_stateless/`. Make some queries, write some data, delete some part (make sure that it stays as outdated for a while). After that ask `asynchronous_metrics` as you did above. You may need to use `query_with_retry` du to asynchronous.

Sure, thank you. This was along the same lines that I was thinking but realized that in the past, it was a bit tricky to add a test that's not flaky. But, it might probably work just fine, I will add it.

bharatnc committed:
Merge upstream/master into ncb/async-metrics-pri-key-size

bharatnc committed:
add a test

CheSema commented:
> I thought, maybe we can make it more lightweight by using CurrentMetrics, with Increments stored inside data parts?

Alexey has a valid point here.
For metrics on top of set with active parts there is no performance penalty. We already iterate on them all for other metrics.

But for my suggestion `TotalConsumptionByPrimaryKeyBytesInMemory` there would be a big penalty if we iterate them all at async metrics job.

Lets put it like best effort. Just check, are you ready to implement `TotalConsumptionByPrimaryKeyBytesInMemory` as CurrentMetrics? Try to do it if you are confident. Or leave it for the next iteration otherwise.

bharatnc commented:
> But for my suggestion `TotalConsumptionByPrimaryKeyBytesInMemory` there would be a big penalty if we iterate them all at async metrics job.


I like your suggestion about `TotalConsumptionByPrimaryKeyBytesInMemory` and I agree that with the current implementation, it doesn't make sense to to include active + inactive parts as it will be expensive.

Which brings us to the question: do we want the metrics `TotalPrimaryKeyBytesInMemory` and `TotalPrimaryKeyBytesInMemoryAllocated` at all ? These two will only track memory used by active parts (I think it would be a nice to have in the asynchronous_metrics).

If we want these metrics, then we can limit the PR to just those (I've also added an integration test for those) - I need to also clarify that detail in the metrics description and docs

On the otherhand, if we only want `TotalConsumptionByPrimaryKeyBytesInMemory` and `TotalConsumptionByPrimaryKeyBytesInMemoryAllocated`, then I can modify this PR to do that instead. 

> Lets put it like best effort. Just check, are you ready to implement `TotalConsumptionByPrimaryKeyBytesInMemory` as CurrentMetrics? Try to do it if you are confident. Or leave it for the next iteration otherwise.

I took a look at CurrentMetrics and it looks like we just need to define these metrics and do the add / sub  for these metrics inside `IMergeTreeDataPart`  when part is loaded and deleted which will inturn expose those in the `system.metrics` table (I might have some clarifications which I can ask later).

bharatnc commented:
> You may need to use `query_with_retry` du to asynchronous.

I actually ended up adding an integration test (I just realized that I have added similar tests long in the past). There I have used `asynchronous_metrics_update_period_s` to be 1s for metrics to be updated and be available faster.  Currently, the tests seem to work fine, I can also additionally replace `query` with `query_with_retry` if recommended.

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,57 @@
+        test_query = "SELECT count() > 0 ? 'ok' : 'fail' FROM system.asynchronous_metric_log WHERE metric='TotalPrimaryKeyBytesInMemoryAllocated';"
+        assert "ok\n" in node1.query(test_query)
+
+    finally:
```
no need to shutdown cluster here. You have a fixture to do this.

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,57 @@
+    try:
+        cluster.start()
+        node1.query("SET log_queries = 1;")
+        node1.query("CREATE DATABASE replica;")
```
you dont need to create db here at all

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,57 @@
+        Engine=MergeTree()
+        PARTITION BY toYYYYMMDD(event_time)
+        ORDER BY id;"""
+        time.sleep(2)
```
no need to sleep.

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,57 @@
+        node1.query("SYSTEM FLUSH LOGS;")
+
+        # query system.asynchronous_metrics
+        test_query = "SELECT count() > 0 ? 'ok' : 'fail' FROM system.asynchronous_metrics WHERE metric='TotalPrimaryKeyBytesInMemory';"
```
I suspect there is a race between query and `asynchronous_metrics_update_period_s`

bharatnc replied:
Interesting, what are you observing?

CheSema replied:
My suspicions is the following:
- async metrics calc new values time to time
-  `SYSTEM FLUSH LOGS` flushes the current values to the logs, they could be stale
To overcome this, you need to do retries with the series `flush + select`.

Also I suspect that the values are not zero even without insertion. Some logs tables could have `MergeTree` strorage, so they all add some value to the sum.
To overcome this, you could do the following:
- ask metric before insertion.
- make table index with granularity 1
- insert 1m rows
- check that you counters have raise accordingly

bharatnc replied:
Ah I see. I actually wrote the tests to be simple and just test if the fields get populated without error and so it tests nothing more.

But I think your suggestion makes sense. I will probably write a test that checks the value before and after and asserts if it's greater after inserting data and lesser after deleting data (couldn't probably assert the exact value, but asserting for > or < would work fine) and I will add retries.

bharatnc replied:
Added the test. `asynchronous_metrics_update_period_s` works for higher values too (for very high values need to override the current default timeout and retries in the test) but setting it to 1 for the purposes of CI (similar to other such integration tests).

CheSema replied:
In general it is all good. I have left some little remarks about code in the test.

Now, when you test is good and survive long `update_period_s`, it is ready to by just `.sh` stateless test.
Stateless tests are chipper. Even when the test is waiting for a minute or two. When stateless test is waiting it consume only one process in CI. A lot of them are run in parallel. Integration test is heavy due to starting new instance of server in separate docker container.
This is optional. I do not insist.

bharatnc replied:
Thank you for your reviews. I've addressed the remarks. I do agree that stateless tests are lighter, but for testing system metrics I often favor doing it via integration tests as the tests are more isolated.

Having said that,  can I try your suggestion as a follow up later? I can work on converting it to a stateless test and open a PR later. It could possibly work well as a stateless test too.

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,57 @@
+import time
```
integration test is good.
I was mentioning the stateless tests, just because they are even more simple than integration.

CheSema commented:
This metrics are valuable for only active parts too.
So you can easily postpone  `TotalConsumptionByPrimaryKeyBytesInMemory` realisation.

bharatnc committed:
fix tests

bharatnc committed:
add a clarification to docs

bharatnc commented:
> This metrics are valuable for only active parts too. So you can easily postpone `TotalConsumptionByPrimaryKeyBytesInMemory` realisation.

Sure that makes sense, I will open a followup PR for the `TotalConsumptionByPrimaryKeyBytesInMemory` implementation.

bharatnc committed:
better test with retries

bharatnc committed:
Merge with upstream/master

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,131 @@
+
+    # alter the table to drop some data
+    node.query(
+        "ALTER table test_pk_bytes DELETE where a < 1000000 SETTINGS mutations_sync=1;"
```
Lest change this query. It is a bad example how to drop all the data in clickhouse.
`truncate table test_pk_bytes` is most efficient.

bharatnc replied:
I'm was dropping some data but not all data here which is why I used ALTER. But, we can surely truncate the table too and I have updated the test.

CheSema commented on tests/integration/test_asynchronous_metrics_pk_bytes_fields/test.py:
```diff
@@ -0,0 +1,131 @@
+    return b < a
+
+
+def query_until_condition(a, b, condition, retries=20, timeout=60, delay=0.5):
```
check `assert_eq_with_retry` from `tests/integration/helpers/test_tools.py`.

bharatnc replied:
Yeah, I checked it out at the time of writing the test, it asserts for equality but I was looking for greater or lesser comparisons, so went this way.

bharatnc committed:
fix test query

bharatnc commented:
test failures don't look related:
- ClickHouse Integration Tests (release) [4/4]  `test_replicated_merge_tree_s3_zero_copy/test.py::test_drop_table`
-  ClickHouse Stateful Tests (debug, ParallelReplicas) `00165_jit_aggregate_functions`
- ClickHouse Upgrade Check (asan)  `Cannot start clickhouse-server`
```
 [ 260418 ] {} <Error> Application: Caught exception while loading metadata: Code: 695. DB::Exception: Load job \'load table test_5.testCleanupR1\' failed: Code: 36. DB::Exception: Wrong number of engine arguments.: Cannot attach table `test_5`.`testCleanupR1` from metadata file /var/lib/clickhouse/st\n [ 260418 ]
 ```

bharatnc commented:
> It does not include projection parts, should we count them as well?

I think that makes sense, opened a PR: https://github.com/ClickHouse/ClickHouse/pull/102978
